File: 임현규/programmers/Lv2/무인도_여행하기/java/UnionFind.java
import java.util.HashMap;
import java.util.Map;
import java.util.stream.IntStream;

class Solution {

    public int[] solution(String[] maps) {
        UnionFind uf = merge(maps);
        Map<Integer, Integer> islands = new HashMap<>();
        for (int i = 0; i < maps.length; ++i) {
            for (int j = 0; j < maps[0].length(); ++j) {
                if (maps[i].charAt(j) == 'X') {
                    continue;
                }
                int node = i * maps[0].length() + j;
                int root = uf.find(node);
                int value = maps[i].charAt(j) - '0';
                islands.put(root, islands.getOrDefault(root, 0) + value);
            }
        }
        int[] result =  islands.values().stream().mapToInt(Integer::intValue).sorted().toArray();
        return result.length == 0 ? new int[]{-1} : result;
    }

    private UnionFind merge(String[] maps) {
        int rows = maps.length;
        int cols = maps[0].length();
        UnionFind uf = new UnionFind(rows * cols);
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (maps[i].charAt(j) == 'X') {
                    continue;
                }
                int node = i * cols + j;
                int rightNode = node + 1;
                int downNode = node + cols;
                if (j + 1 < cols && maps[i].charAt(j + 1) != 'X') {
                    uf.merge(node, rightNode);
                }
                if (i + 1 < rows && maps[i + 1].charAt(j) != 'X') {
                    uf.merge(node, downNode);
                }
            }
        }
        return uf;
    }

    static class UnionFind {

        private final int[] parent;

        public UnionFind(int size) {
            this.parent = IntStream.range(0, size).toArray();
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        public void merge(int x, int y) {
            x = find(x);
            y = find(y);
            if (x != y) {
                parent[y] = x;
            }
        }
    }
}

File: 임현규/programmers/Lv2/무인도_여행하기/python/unionfind.py
import sys
from collections import defaultdict

sys.setrecursionlimit(10 ** 7)


class UnionFind:

    def __init__(self, n):
        self.parents = [i for i in range(n)]

    def find(self, x):
        if self.parents[x] != x:
            self.parents[x] = self.find(self.parents[x])
        return self.parents[x]

    def union(self, x, y):
        parent_x = self.find(x)
        parent_y = self.find(y)
        self.parents[parent_y] = parent_x

    def is_connect(self, x, y):
        return self.find(x) == self.find(y)


def solution(maps):
    rows = len(maps)
    cols = len(maps[0])

    def is_in_map(x, y):
        return 0 <= x < rows and 0 <= y < cols

    def is_number(x, y):
        return maps[x][y] != 'X'

    uf = UnionFind(rows * cols)
    for i in range(rows):
        for j in range(cols):
            node = i * cols + j
            right_node = i * cols + j + 1
            down_node = (i + 1) * cols + j
            if is_in_map(i, j + 1) and is_number(i, j + 1):
                uf.union(node, right_node)
            if is_in_map(i + 1, j) and is_number(i + 1, j):
                uf.union(node, down_node)

    islands = defaultdict(int)
    for i in range(rows):
        for j in range(cols):
            if is_number(i, j):
                islands[uf.find(i * cols + j)] += int(maps[i][j])
    return sorted(islands.values()) if islands else [-1]